### ISSUE TITLE
Bishoujo Senshi Sailormoon Sailorstars - Fuwafuwa Panic

### ISSUE THREAD

**Issue by zack867** [bsnes-emu/bsnes#323]
in bsnes v115-windows, 
when you load Bishoujo Senshi Sailormoon Sailorstars - Fuwafuwa Panic.
before loading the game, bsnes will ask you to load sufami turbo rom for slot A and B first like you are loading empty sufami turbo system which is weird.
because Bishoujo Senshi Sailormoon Sailorstars - Fuwafuwa Panic is not a sufami turbo game.
instead, Bishoujo Senshi Sailormoon Sailorstars - Fuwafuwa Panic 2 is a sufami turbo game.
this didn't happen in bsnes v073.

**Comment by Screwtapello** [comment]
As far as I can tell, there's no game called *Bishoujo Senshi Sailormoon Sailorstars - Fuwafuwa Panic*, but there is a game called [*Bishoujo Senshi Sailor Moon SuperS - Fuwafuwa Panic*](https://datomatic.no-intro.org/index.php?page=show_record&s=49&n=0250), which I assume is what you're talking about. No-Intro includes [a picture of the game's PCB](https://datomatic.no-intro.org/attachments%20-%20sources/Nintendo%20-%20Super%20Nintendo%20Entertainment%20System/20221215_084342_f2.JPG), confirming that it is does *not* have slots for accepting Sufami Turbo games, and therefore bsnes should not prompt for them to be inserted.

(*Bishoujo Senshi Sailormoon Sailor Stars - Fuwafuwa Panic 2* is a Sufami Turbo game in the sense that it fits into a slot on the Sufami Turbo, so it should *also* not prompt for additional things to be put in slots, but that's not what this issue is about)

Near didn't get around to making an official dump of this game in the board database, so bsnes uses its heuristics to detect the game's memory map as:

```
game
  sha256:   af857c18e168d30b35876aa319171ee363bc262c9d0627176573e8311d43297d
  label:    Bishoujo Senshi Sailor Moon SuperS - Fuwafuwa Panic (Japan)
  name:     Bishoujo Senshi Sailor Moon SuperS - Fuwafuwa Panic (Japan)
  title:    SFC SAILORMOON SUPERS
  region:   NTSC
  revision: SHVC-A9PJ-0
  board:    ST-LOROM
    memory
      type: ROM
      size: 0x100000
      content: Program
```

"ST-LOROM" is apparently a generic "board with Sufami Turbo slots" memory map; the actual Sufami Turbo's board is "BANDAI-PT-923". Looking at bsnes' heuristic detection, it seems that a LOROM board is converted into an ST-LOROM board if the serial in the ROM header is "A9PJ":

https://github.com/bsnes-emu/bsnes/blob/79770f6207a5d244652644c5e2d9f4328bd974ac/bsnes/heuristics/super-famicom.cpp#L277-L279

As we can see in the detected memory map above, that is indeed the serial of *Fuwafuwa Panic*. According to Near's official dump, that is also the serial in the Sufami Turbo base cart:

https://github.com/bsnes-emu/bsnes/blob/79770f6207a5d244652644c5e2d9f4328bd974ac/bsnes/Database/Super%20Famicom.bml#L1212-L1222

This greatly surprises me, I thought Nintendo were pretty good about making sure serial numbers were unique. Possibly this happened because the Sufami Turbo was manufactured by Bandai instead of Nintendo?

I suspect the easiest fix is to just remove the special-case handling of serial A9PJ, since we have an Official Dump of the Sufami Turbo. On the other hand, that would prevent homebrew devs from making custom games that take Sufami Turbo mini-carts, much like how there are various games that take Satellaview mini-carts. I'm not sure there's anything else in the game header that would distinguish "ROM that wants to be a Sufami Turbo base cart" from any other ROM, except the serial and (perhaps) the size.

It may be worth trying to extract the ROM headers of the SNES library to figure out what an acceptable heuristic might be.

**Comment by Screwtapello** [comment]
ares inherited the same heuristics that bsnes has.

MesenS doesn't mention "A9PJ" or "Sufami" in its source, I assume it doesn't emulate the Sufami Turbo at all.

snes9x detects the Sufami Turbo base cart by its file size and the strings "BANDAI SFC-ADX" and "SFC-ADX BACKUP", which appear at the very beginning of the ROM dump. Those would be very nice fingerprints for custom ROMs to use, and custom ROMs would need to use them anyway to be compatible with snes9x.

**Comment by Screwtapello** [comment]
For the sake of consistency, we should probably fix this the same way ares does, so I've made a PR over there:

https://github.com/ares-emulator/ares/pull/1628

Once that's applied, I'll make the same change for bsnes.

**Comment by Screwtapello** [comment]
Thanks for reporting this issue! This should be fixed in the latest nightly build.

**Comment by Screwtapello** [comment]
Apparently the change that fixed this issue causes a crash on Windows. I guess we need to dig deeper.

**Comment by Screwtapello** [comment]
Another attempt at a fix in #326.

**Comment by Screwtapello** [comment]
OK, *now* it should be fixed.
